v2/apiserver/internal/core/kubernetes/substrate.go

```diff
@@ -52,6 +83,84 @@ func NewSubstrate(
 	}
 }
 
+func (s *substrate) CountRunningWorkers(
+	ctx context.Context,
+) (core.SubstrateWorkerCount, error) {
+	count := core.SubstrateWorkerCount{}
+	phaseSelectors := []string{
+		pendingPodsSelector,
+		runningPodsSelector,
+		unknownPhasePodsSelector,
+	}
+	// Use a set to assist in counting. This helps prevent us from double-counting
+	// a pod if its phase changes from pending to running while we're counting.
+	podsSet := map[string]struct{}{}
+	for _, phaseSelector := range phaseSelectors {
+		var cont string
+		for {
+			pods, err := s.kubeClient.CoreV1().Pods("").List(
+				ctx,
+				metav1.ListOptions{
+					LabelSelector: workerPodsSelector,
+					FieldSelector: phaseSelector,
+					Continue:      cont,
+				},
+			)
+			if err != nil {
+				return count, errors.Wrap(err, "error counting running worker pods")
+			}
+			for _, pod := range pods.Items {
+				podsSet[fmt.Sprintf("%s:%s", pod.Namespace, pod.Name)] = struct{}{}
+			}
+			cont = pods.Continue
+			if cont == "" {
+				break
+			}
+		}
+	}
+	count.Count = len(podsSet)
+	return count, nil
+}
+
+func (s *substrate) CountRunningJobs(
+	ctx context.Context,
+) (core.SubstrateJobCount, error) {
+	count := core.SubstrateJobCount{}
+	phaseSelectors := []string{
+		pendingPodsSelector,
+		runningPodsSelector,
+		unknownPhasePodsSelector,
+	}
+	// Use a set to assist in counting. This helps prevent us from double-counting
+	// a pod if its phase changes from pending to running while we're counting.
+	podsSet := map[string]struct{}{}
+	for _, phaseSelector := range phaseSelectors {
```

Comment:
As far as I can tell, the two functions here share almost all the logic for actually counting pods - the only difference seems to be the pod selector passed to the pod client.

Would it make sense to avoid duplicating?

Reply:
Yes. I definitely overlooked that. Fixing it now.

Reply:
Fixed in recent force push.